app/util/statusPoll.js

```diff
@@ -0,0 +1,98 @@
+const serviceState = {
+  UP: Symbol('status-up'),
+  DOWN: Symbol('status-down'),
+  ERROR: Symbol('status-error'),
+}
+const UNKNOWN = Symbol('status-unknown')
+const stateSymbols = new Set(Object.values(serviceState))
+
+const startStatusHandler = async ({ pollingPeriodMs, serviceTimeoutMs, getStatus }) => {
+  const status = {
+    status: UNKNOWN,
+    detail: null,
+  }
+  let timeoutHandle = null,
+    stop = false
+
+  const resetStatus = () => {
+    status.status = serviceState.ERROR
+    status.detail = null
+  }
+
+  const updateStatus = async () => {
+    try {
+      const newStatus = await Promise.race([
+        getStatus(),
+        new Promise((resolve) => setTimeout(resolve, serviceTimeoutMs, { status: serviceState.ERROR, detail: null })),
+      ])
+
+      if (stateSymbols.has(newStatus.status)) {
+        status.status = newStatus.status
+        status.detail = newStatus.detail === undefined ? null : newStatus.detail
+      } else {
+        resetStatus()
+      }
+    } catch (err) {
+      resetStatus()
+    }
+  }
+
+  const statusLoop = async () => {
```

Comment:
found this a little confusing, but hard to think of a better solution so far I got a couple of things in my mind so it's either the generator or iterator, both going to the infinity with some timeout.

Reply:
generator might be better. I'll have a look

Reply:
updated to a generator. I think it's a bit cleaner now @n3op2